Custom stash capacity per entity (EventSourced and Durable)

sebastian-alfers opened the pull request:
Fixes https://github.com/akka/akka/issues/31478.

Not sure why but when I ran `applyCodeStyle` some files were changed that I did not touch.

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/EventSourcedBehavior.scala:
```diff
@@ -232,6 +240,8 @@ abstract class EventSourcedBehavior[Command, Event, State] private[akka] (
     scaladsl.EventSourcedBehavior.lastSequenceNumber(ctx.asScala)
   }
 
+  def withStashCapacity: Optional[java.lang.Integer] = Optional.empty()
```
Add JavaDoc, something starting with "Override and define..."

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/EventSourcedBehavior.scala:
```diff
@@ -232,4 +232,6 @@ object EventSourcedBehavior {
   @InternalApi
   private[akka] def withReplication(context: ReplicationContextImpl): EventSourcedBehavior[Command, Event, State]
+
+  def withStashCapacity(size: Int): EventSourcedBehavior[Command, Event, State]
```
doc

patriknw reviewed (commented):
looking good after adding some more docs

sebastian-alfers commented:
@patriknw I added the missing JavaDocs

johanandren reviewed (approved):
LGTM

johanandren commented:
@sebastian-alfers it is better to rebase rather than merge master, since it allows us to do a single squashed commit when PR is done. If you can drop the merge in an interactive rebase on top of current main and force push I think that will sort things.

sebastian-alfers committed:
wip

override stash-capacity per entity (EventSourced and Durable)

small fixes

indention

indention

newline

fmt

mima + small changes

added some JavaDocs

sebastian-alfers commented:
Thanks @johanandren for pointing that out. I managed to rebase and resolve the conflicts.

johanandren commented:
Aha, it actually does need some MiMa excludes for those new methods

sebastian-alfers committed:
mima filters against 2.6.20

johanandren reviewed (approved):
LGTM!

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/EventSourcedBehavior.scala:
```diff
@@ -234,6 +242,12 @@ abstract class EventSourcedBehavior[Command, Event, State] private[akka] (
+   * Override to define a custom stash capacity per entity.
+   * If not defined, the default `akka.persistence.typed.stash-capacity` will be used.
+   */
+  def withStashCapacity: Optional[java.lang.Integer] = Optional.empty()
```
To align better with the naming of the other methods that can be overridden here I think we should name this `stashCapacity`.


scaladsl is different because there it's a "builder"

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/state/scaladsl/DurableStateBehavior.scala:
```diff
@@ -160,4 +160,9 @@ object DurableStateBehavior {
+   * Override to define a custom stash capacity per entity.
+   * If not defined, the default `akka.persistence.typed.stash-capacity` will be used.
+   */
+  def withStashCapacity(size: Int): DurableStateBehavior[Command, State]
```
missing corresponding in javadsl.DurableStateBehavior?

patriknw reviewed (commented):
looking good, just something small...

sebastian-alfers committed:
rename withStashCapacity in the java api, add stashCapacity for the java api for Durable State

sebastian-alfers committed:
code style

octonato commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/EventSourcedBehavior.scala:
```diff
@@ -221,10 +221,18 @@ abstract class EventSourcedBehavior[Command, Event, State] private[akka] (
+      else
+        behaviorWithSignalHandler
+
+    if (stashCapacity.isPresent()) {
```
```suggestion
    if (stashCapacity.isPresent) {
```    


I tried to use `map/orElse` for this, but surprisingly type inference is not working as expected. 🤷

octonato replied:
This is not an important suggestion. Can skip if you prefer. 

I added here because intellij gave the suggestion. I'm just passing it over. 😄

octonato commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/EventSourcedBehavior.scala:
```diff
@@ -232,4 +232,10 @@ object EventSourcedBehavior {
   private[akka] def withReplication(context: ReplicationContextImpl): EventSourcedBehavior[Command, Event, State]
+
+  /**
+   * Override to define a custom stash capacity per entity.
+   * If not defined, the default `akka.persistence.typed.stash-capacity` will be used.
+   */
```
This trait is marked with `@DoNotInherit`. Users of the scaladsl don't extend this and therefore won't have the change to override this method. Instead, they get an instance of `EventSourcedBehaviorImpl` to configure further.

octonato commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/state/javadsl/DurableStateBehavior.scala:
```diff
@@ -136,10 +136,17 @@ abstract class DurableStateBehavior[Command, State] private[akka] (
+      else
+        behaviorWithSignalHandler
+
+    if (stashCapacity.isPresent()) {
```
```suggestion
    if (stashCapacity.isPresent) {
```

same as above

octonato commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/state/scaladsl/DurableStateBehavior.scala:
```diff
@@ -160,4 +160,9 @@ object DurableStateBehavior {
   def onPersistFailure(backoffStrategy: BackoffSupervisorStrategy): DurableStateBehavior[Command, State]
 
+  /**
+   * Override to define a custom stash capacity per entity.
```
Same as above. No Overriding in the scaladsl.
This trait is marked with `@DoNotInherit`.

sebastian-alfers replied:
That makes sense. I changed for both.

octonato reviewed (approved):
LGTM, after some minor change on scaladoc.

sebastian-alfers committed:
Update akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/EventSourcedBehavior.scala

Co-authored-by: Renato Cavalcanti <renato@cavalcanti.be>

sebastian-alfers committed:
Update akka-persistence-typed/src/main/scala/akka/persistence/typed/state/javadsl/DurableStateBehavior.scala

Co-authored-by: Renato Cavalcanti <renato@cavalcanti.be>

sebastian-alfers committed:
adjust comment for javadsl to indicate that no override is involved

patriknw reviewed (approved):
LGTM
